File: car/tcp_client.py
import socket
import time
import cv2
import numpy as np
import threading
import argparse
import ipaddress
import sys

(major_ver, minor_ver, subminor_ver) = (cv2.__version__).split('.')
print(cv2.__version__)

tracker_types = ['BOOSTING', 'MIL','KCF', 'TLD', 'MEDIANFLOW', 'CSRT', 'MOSSE']
tracker_type = tracker_types[5]

parser = argparse.ArgumentParser(description='Connect to raspberry pi using IP address')
parser.add_argument('host', metavar='h', type=str, help='ip address of pi')
args = parser.parse_args()

def init_tracker(frame, bbox):
    if int(major_ver) < 4 and int(minor_ver) < 3:
        tracker = cv2.cv2.Tracker_create(tracker_type)
    else:
        if tracker_type == 'BOOSTING':
            tracker = cv2.TrackerBoosting_create()
        if tracker_type == 'MIL':
            tracker = cv2.TrackerMIL_create()
        if tracker_type == 'KCF':
            tracker = cv2.TrackerKCF_create()
        if tracker_type == 'TLD':
            tracker = cv2.TrackerTLD_create()
        if tracker_type == 'MEDIANFLOW':
            tracker = cv2.TrackerMedianFlow_create()
        if tracker_type == 'CSRT':
            tracker = cv2.TrackerCSRT_create()
        if tracker_type == 'MOSSE':
            tracker = cv2.legacy.TrackerMOSSE_create()
    tracker.init(frame, bbox)
    return tracker

class CameraBufferCleanerThread(threading.Thread):
    def __init__(self, camera, name='camera-buffer-cleaner-thread'):
        self.camera = camera
        self.last_frame = None
        self.running = True
        super(CameraBufferCleanerThread, self).__init__(name=name)
        self.start()

    def run(self):
        while self.running:
            _, self.last_frame = self.camera.read()
    
def get_decision(bbox):
    x, w = int(bbox[0]), int(bbox[2])
    cx = int(x + w // 2)

    # Calculate slope
    if cx < width // 2 - 50: 
        return "Turn left"
    elif cx > width // 2 + 50: 
        return "Turn right"
    else:  
        return "Go straight"

def count_red_pixels(image):
    # Convert image to HSV color space
    hsv = cv2.cvtColor(image, cv2.COLOR_BGR2HSV)
    h, s, v = cv2.split(hsv)
    v_eq = cv2.equalizeHist(v)
    hsv = cv2.merge([h, s, v_eq])
    
    # Define lower and upper bounds for red color
    lower_red = np.array([0, 100, 100])
    upper_red = np.array([10, 255, 255])

    # Mask out red pixels
    mask1 = cv2.inRange(hsv, lower_red, upper_red)

    # Define lower and upper bounds for slightly different red color
    lower_red2 = np.array([160, 100, 100])
    upper_red2 = np.array([180, 255, 255])

    # Mask out the second range of red pixels
    mask2 = cv2.inRange(hsv, lower_red2, upper_red2)

    # Combine masks
    mask = cv2.bitwise_or(mask1, mask2)

    # Count red pixels
    red_pixel_count = cv2.countNonZero(mask)

    return red_pixel_count

# Define host and port
try:
    HOST = str(ipaddress.ip_address(args.host))
    print(f"Valid IP address: {HOST}")
except ValueError:
    sys.exit(f"Invalid IP address: {args.host}")
PORT = 65432

cap = cv2.VideoCapture(f"rtsp://{HOST}:8554/video_stream")
cam_cleaner = CameraBufferCleanerThread(cap)

net = cv2.dnn.readNetFromONNX("best1404.onnx")
classes = ['obstacle', 'parking']

bbox = None
tracking = False
obj_label = ""
conf = -1

decision = ""

threshold = 120000

# Create a socket object
with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as client_socket:
    # Connect to server
    client_socket.connect((HOST, PORT))
    
    # Main loop
    while True and decision != "Stop":
        # Read frame from webcam
        if cam_cleaner.last_frame is None:
            continue

        frame = cam_cleaner.last_frame

        # Start timer
        timer = cv2.getTickCount()

        if tracking:
            ok, bbox = tracker.update(frame)
            # Draw bounding box
            if ok:
                x, y, w, h = int(bbox[0]), int(bbox[1]), int(bbox[2]), int(bbox[3])
                cx = int(x + w // 2)
                cy = int(y + h // 2)
                cv2.line(frame, (int(cx), int(cy)), (width // 2, height), (255,255,255), 2)

                decision = get_decision(bbox) 
                
                num_red_pixel = count_red_pixels(frame)
                if num_red_pixel > threshold:
                    decision = "Stop"
                    tracking = False
                else:
                    print(num_red_pixel)
                    
                client_socket.sendall(decision.encode())
                text = obj_label + "{:.2f}".format(conf)
                cv2.rectangle(frame, (x, y), (x + w, y + h), (255,0,0), 2)
                cv2.putText(frame, text, (x,y-2),cv2.FONT_HERSHEY_COMPLEX, 0.7,(255,0,255),2)
                cv2.putText(frame, decision, (160, 80), cv2.FONT_HERSHEY_SIMPLEX, 2, (0, 0, 255), 2)  
            else :
                tracking = False
        else: 
            # Detect objects using YOLOv5
            height, width, _ = frame.shape
            blob = cv2.dnn.blobFromImage(frame, 1/255, (640, 640), (0, 0, 0), True, crop=False)
            net.setInput(blob)
            detections = net.forward()[0]
            
            classes_ids = []
            confidences = []
            boxes = []
            rows = detections.shape[0]

            x_scale = width/640
            y_scale = height/640

            for i in range(rows):
                row = detections[i]
                confidence = row[4]
                if confidence >= 0.4:
                    classes_score = row[5:]
                    ind = np.argmax(classes_score)
                    if classes_score[ind] >= 0.25:
                        classes_ids.append(ind)
                        confidences.append(confidence)
                        cx, cy, w, h = row[:4]
                        x1 = int((cx- w/2)*x_scale)
                        y1 = int((cy-h/2)*y_scale)
                        wv= int(w * x_scale)
                        hv = int(h * y_scale)
                        box = np.array([x1,y1,wv,hv])
                        boxes.append(box)

            if len(boxes) > 0:
                num_retained_boxes = cv2.dnn.NMSBoxes(boxes,confidences,0.25,0.45)
                parking_id = -1
                max_conf = -1
                for i in num_retained_boxes:
                    if classes[classes_ids[i]] == 'parking' and max_conf < confidences[i]:
                        parking_id = i
                        max_conf = confidences[i]
                    
                if parking_id > -1:
                    bbox = boxes[parking_id]
                    x1,y1,w,h = bbox
                    obj_label = classes[classes_ids[parking_id]]
                    conf = max_conf
                    text = obj_label + "{:.2f}".format(conf)
                    cv2.rectangle(frame,(x1,y1),(x1+w,y1+h),(255,0,0),2)
                    cv2.putText(frame, text, (x1,y1-2),cv2.FONT_HERSHEY_COMPLEX, 0.7,(255,0,255),2)
                    cv2.putText(frame, decision, (20, 80), cv2.FONT_HERSHEY_SIMPLEX, 2, (0, 0, 255), 2)
                    tracker = init_tracker(frame, bbox)
                    tracking = True
                    client_socket.sendall(get_decision(bbox).encode())
                        
            
            if not tracking:
                client_socket.sendall("None".encode())
        
        # Calculate Frames per second (FPS)
        fps = cv2.getTickFrequency() / (cv2.getTickCount() - timer)
        # Display FPS on frame
        cv2.putText(frame, "FPS : " + str(int(fps)), (20,50), cv2.FONT_HERSHEY_SIMPLEX, 0.75, (50,170,50), 2)
        
        # Display frame with bounding box
        cv2.imshow("Frame", frame)

        if cv2.waitKey(1) & 0xFF == ord('q'): 
            break 
    cam_cleaner.running = False
    time.sleep(2)
    cap.release()
    client_socket.close()    

File: car/tcp_client.cpp
#include <opencv2/opencv.hpp>
#include <opencv2/tracking/tracking.hpp>
#include <string>
#include <fstream>
#include <cstring>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>

std::vector<std::string> load_class_list()
{
    std::vector<std::string> class_list;
    std::ifstream ifs("config_files/classes.txt");
    std::string line;
    while (getline(ifs, line))
    {
        class_list.push_back(line);
    }
    return class_list;
}

void load_net(cv::dnn::Net &net, bool is_cuda)
{
    auto result = cv::dnn::readNet("config_files/best1404.onnx");
    if (is_cuda)
    {
        std::cout << "Attempty to use CUDA\n";
        result.setPreferableBackend(cv::dnn::DNN_BACKEND_CUDA);
        result.setPreferableTarget(cv::dnn::DNN_TARGET_CUDA_FP16);
    }
    else
    {
        std::cout << "Running on CPU\n";
        result.setPreferableBackend(cv::dnn::DNN_BACKEND_OPENCV);
        result.setPreferableTarget(cv::dnn::DNN_TARGET_CPU);
    }
    net = result;
}

cv::Ptr<cv::Tracker> init_tracker(cv::Mat frame, cv::Rect2d bbox) {
    cv::Ptr<cv::Tracker> tracker = cv::TrackerCSRT::create();
    tracker->init(frame, bbox);
    return tracker;
}

std::string get_decision(cv::Rect bbox) {
    return "Decision";
}

const std::vector<cv::Scalar> colors = {cv::Scalar(255, 255, 0), cv::Scalar(0, 255, 0), cv::Scalar(0, 255, 255), cv::Scalar(255, 0, 0)};

const float INPUT_WIDTH = 640.0;
const float INPUT_HEIGHT = 640.0;
const float SCORE_THRESHOLD = 0.2;
const float NMS_THRESHOLD = 0.4;
const float CONFIDENCE_THRESHOLD = 0.4;

struct Detection
{
    int class_id;
    float confidence;
    cv::Rect box;
};

cv::Mat format_yolov5(const cv::Mat &source) {
    int col = source.cols;
    int row = source.rows;
    int _max = MAX(col, row);
    cv::Mat result = cv::Mat::zeros(_max, _max, CV_8UC3);
    source.copyTo(result(cv::Rect(0, 0, col, row)));
    return result;
}

void detect(cv::Mat &image, cv::dnn::Net &net, std::vector<Detection> &output, const std::vector<std::string> &className) {
    cv::Mat blob;

    auto input_image = format_yolov5(image);
    
    cv::dnn::blobFromImage(input_image, blob, 1./255., cv::Size(INPUT_WIDTH, INPUT_HEIGHT), cv::Scalar(), true, false);
    net.setInput(blob);
    std::vector<cv::Mat> outputs;
    net.forward(outputs, net.getUnconnectedOutLayersNames());

    float x_factor = input_image.cols / INPUT_WIDTH;
    float y_factor = input_image.rows / INPUT_HEIGHT;
    
    float *data = (float *)outputs[0].data;

    const int dimensions = 85;
    const int rows = 25200;
    
    std::vector<int> class_ids;
    std::vector<float> confidences;
    std::vector<cv::Rect> boxes;

    for (int i = 0; i < rows; ++i) {

        float confidence = data[4];
        if (confidence >= CONFIDENCE_THRESHOLD) {

            float * classes_scores = data + 5;
            cv::Mat scores(1, className.size(), CV_32FC1, classes_scores);
            cv::Point class_id;
            double max_class_score;
            minMaxLoc(scores, 0, &max_class_score, 0, &class_id);
            if (max_class_score > SCORE_THRESHOLD) {

                confidences.push_back(confidence);

                class_ids.push_back(class_id.x);

                float x = data[0];
                float y = data[1];
                float w = data[2];
                float h = data[3];
                int left = int((x - 0.5 * w) * x_factor);
                int top = int((y - 0.5 * h) * y_factor);
                int width = int(w * x_factor);
                int height = int(h * y_factor);
                boxes.push_back(cv::Rect(left, top, width, height));
            }

        }

        data += 85;
    }

    std::vector<int> nms_result;
    cv::dnn::NMSBoxes(boxes, confidences, SCORE_THRESHOLD, NMS_THRESHOLD, nms_result);
    for (int i = 0; i < nms_result.size(); i++) {
        int idx = nms_result[i];
        Detection result;
        result.class_id = class_ids[idx];
        result.confidence = confidences[idx];
        result.box = boxes[idx];
        output.push_back(result);
    }
}

int main(int argc, char**argv)
{
    std::cout << "Enter pi's ip address: ";
    std::string server_ip{};
    std::getline(std::cin, server_ip);
    int port{65432};
    std::vector<std::string> class_list = load_class_list();

    bool is_cuda = argc > 1 && strcmp(argv[1], "cuda") == 0;
    
    cv::dnn::Net net;
    load_net(net, is_cuda);

    int sockfd = socket(AF_INET, SOCK_STREAM, 0);
    if (sockfd < 0) {
        std::cerr << "Error creating socket" << std::endl;
        return 1;
    }

    // Server address structure
    struct sockaddr_in serverAddr;
    memset(&serverAddr, 0, sizeof(serverAddr));
    serverAddr.sin_family = AF_INET;
    serverAddr.sin_port = htons(8080); // Port number

    // Convert IP address from presentation format to network format
    if (inet_pton(AF_INET, server_ip.c_str(), &serverAddr.sin_addr) <= 0) {
        std::cerr << "Invalid address/ Address not supported" << std::endl;
        return 1;
    }

    // Connect to the server
    // if (connect(sockfd, (const sockaddr*)&serverAddr, sizeof(serverAddr)) < 0) {
    //     std::cerr << "Connection failed" << std::endl;
    //     return 1;
    // }

    cv::Mat frame;
    cv::VideoCapture cap("rtsp://" + server_ip + ":8554/video_stream?dummy=test.mjpg", cv::CAP_FFMPEG);
    
    bool is_tracking = false;
    cv::Ptr<cv::Tracker> tracker{};
    cv::Rect bbox{};
    std::string decision{"None"};
    std::string obj_label{};
    int conf{-1};

    while (true)
    {
        cap >> frame;
        if (frame.empty()) {
            continue; // to be replaced with cam cleaner later
        }

        cv::imshow("frame", frame);
        continue;
        if (is_tracking) 
        {
            if (tracker->update(frame, bbox)) 
            {
                int cx = int(bbox.x + bbox.width / 2);
                int cy = int(bbox.y + bbox.height / 2);
                cv::line(frame, cv::Point(cx, cy), cv::Point(INPUT_WIDTH / 2, INPUT_HEIGHT), (255,255,255), 2);

                decision = get_decision(bbox);
                
                // if (send(sockfd, decision.c_str(), decision.length(), 0) < 0) {
                //     std::cerr << "Error sending data" << std::endl;
                //     return 1;
                // }

                std::string text = obj_label;
                cv::rectangle(frame, cv::Point(bbox.x, bbox.y), cv::Point(bbox.x + bbox.width, bbox.y + bbox.height), (255,0,0), 2);
                cv::putText(frame, text, cv::Point(bbox.x,bbox.y-2),cv::FONT_HERSHEY_COMPLEX, 0.7,(255,0,255),2);
                cv::putText(frame, decision, cv::Point(160, 80), cv::FONT_HERSHEY_SIMPLEX, 2, (0, 0, 255), 2); 
            } else {
                is_tracking = false;
            }
        }
        std::vector<Detection> output;
        detect(frame, net, output, class_list);
        
        int detections = output.size();

        for (int i = 0; i < detections; ++i)
        {
            auto detection = output[i];
            auto bbox = detection.box;
            auto classId = detection.class_id;
            if (class_list[detection.class_id] == "parking")
            {
                const auto color = colors[classId % colors.size()];
                // add overlay and init tracking
                tracker = init_tracker(frame, bbox);
                is_tracking = true;
                break;
            }
        }

        cv::imshow("output", frame);

        if (cv::waitKey(1) != -1)
        {
            cap.release();
            std::cout << "finished by user\n";
            break;
        }
    }

    // close(sockfd);
    
    return 0;
}
